modules/cdp_engine/src/lib.rs

```diff
@@ -207,10 +249,126 @@ decl_module! {
 				}
 			}
 		}
+
+		// Call by unsigned transaction
+		pub fn liquidate(
+			origin,
+			currency_id: CurrencyIdOf<T>,
+			who: T::AccountId,
+		) {
+			ensure_none(origin)?;
+			ensure!(!Self::is_shutdown(), Error::<T>::AlreadyShutdown);
+			Self::liquidate_unsafe_cdp(who.clone(), currency_id)?;
+		}
+
+		// Runs after every block.
+		fn offchain_worker(now: T::BlockNumber) {
+			debug::RuntimeLogger::init();
+
+			if let Err(e) = Self::automatic_liquidation() {
+				debug::debug!(
+					target: "cdp-engine offchain worker",
+					"do not start automatic liquidation at {:?}: {:?}",
+					now,
+					e,
+				);
+			}
+		}
 	}
 }
 
 impl<T: Trait> Module<T> {
+	fn automatic_liquidation() -> Result<(), OffchainErr> {
+		// check if is validator
+		if !runtime_io::offchain::is_validator() {
+			return Err(OffchainErr::NotValidator);
+		}
+
+		// check if require offchain worker lock succecced
+		let local_storage_key = DB_PREFIX.to_vec();
+		let local_storage = StorageValueRef::local(&local_storage_key);
+		let collateral_currency_ids = T::CollateralCurrencyIds::get();
+		let mut execute_position: u32 = 0;
+
+		let require_lock = local_storage.mutate(|lock: Option<Option<OffchainWorkerLock>>| {
+			match lock {
+				None => {
+					// start with random collateral
+					let random_seed = runtime_io::offchain::random_seed();
+					let mut rng = RandomNumberGenerator::<BlakeTwo256>::new(BlakeTwo256::hash(&random_seed[..]));
+					execute_position = rng.pick_u32(collateral_currency_ids.len() as u32);
+
+					Ok(OffchainWorkerLock {
+						previous_position: execute_position,
+						locked: true,
+					})
+				}
+				Some(Some(lock)) if !lock.locked => {
+					if lock.previous_position < collateral_currency_ids.len() as u32 - 1 {
+						execute_position = lock.previous_position + 1
+					}
+
+					Ok(OffchainWorkerLock {
+						previous_position: execute_position,
+						locked: true,
+					})
+				}
+				_ => Err(OffchainErr::FailedToAcquireLock),
+			}
+		})?;
+		let _ = require_lock.map_err(|_| OffchainErr::FailedToAcquireLock)?;
```

Comment:
Can we do something like this to avoid `mut` execute_position. Also finding the current flow bit hard to understand.
`let OffchainWorkerLock { previous_position } = require_lock.map_err(|_| OffchainErr::FailedToAcquireLock)?;`